components/ai_chat/resources/untrusted_conversation_frame/components/assistant_response/memory_tool_event.tsx

```diff
@@ -0,0 +1,145 @@
+    try {
+      const input = JSON.parse(toolUseEvent.argumentsJson)
+      return input?.memory || ''
+    } catch (e) {
+      return ''
+    }
+  }, [toolUseEvent?.argumentsJson])
+
+  // Check if memory exists using HasMemory API and subscribe to changes
+  React.useEffect(() => {
+    if (!memoryContent) return
+
+    // Initial check
+    context.uiHandler?.hasMemory(memoryContent).then(({ exists }) => {
+      setMemoryExists(exists)
+    })
+
+    // Subscribe to memory changes via UI observer
+    const id = context.uiObserver?.onMemoriesChanged.addListener(
+      (memories: string[]) => {
+        const exists = memories.includes(memoryContent)
+        setMemoryExists(exists)
+      }
+    )
+
+    return () => {
+      context.uiObserver?.removeListener(id)
+    }
+  }, [memoryContent])
+
+  // Non-empty string in textContentBlock indicates an error
+  const hasError = !!toolUseEvent.output?.[0]?.textContentBlock?.text
+
+  const handleUndo = async () => {
+    if (!memoryContent) return
+
+    await context.uiHandler?.deleteMemory(memoryContent)
+  }
+
+  const handleManageAll = () => {
+    context.uiHandler?.openAIChatCustomizationSettings?.()
+  }
+
+  // Don't render if no memory content, tool hasn't completed, or still
+  // checking memory existence initially.
+  if (!memoryContent || !toolUseEvent.output || memoryExists === null) {
+    return null
+  }
+
+  // Handle error state
+  if (hasError) {
+    return (
+      <div
+        className={styles.memoryToolEvent}
+        data-testid='memory-tool-event-error'
+      >
+        <Icon name='database' className={styles.icon} />
+        <span className={styles.textError}>
+          {getLocale(S.CHAT_UI_MEMORY_ERROR_LABEL)}
+          <button
+            className={`${styles.button} ${styles.buttonWithSpacing}`}
+            onClick={handleManageAll}
+            data-testid='memory-manage-button'
+          >
+            {getLocale(S.CHAT_UI_MEMORY_MANAGE_ALL_BUTTON_LABEL)}
+          </button>
+        </span>
+      </div>
+    )
+  }
+
+  // Handle undone state
+  if (!memoryExists) {
+    return (
+      <div
+        className={styles.memoryToolEvent}
+        data-testid='memory-tool-event-undone'
+      >
+        <Icon name='database' className={styles.icon} />
+        <span className={styles.textUndone} title={memoryContent}>
+          {getLocale(S.CHAT_UI_MEMORY_UNDONE_LABEL)}
+          <button
+            className={`${styles.button} ${styles.buttonWithSpacing}`}
+            onClick={handleManageAll}
+            data-testid='memory-manage-button'
+          >
+            {getLocale(S.CHAT_UI_MEMORY_MANAGE_ALL_BUTTON_LABEL)}
+          </button>
+        </span>
+      </div>
+    )
+  }
+
+  return (
+    <div className={styles.memoryToolEvent} data-testid='memory-tool-event'>
+      <Icon name='database' className={styles.icon} />
+      <span>
+        {getLocale(S.CHAT_UI_MEMORY_UPDATED_LABEL)} {memoryContent}{' '}
```

Comment:
In all languages will it be the same order? I'm not sure, or if there's a suffix in some languages. Normally this should use a placeholder and `formatLocale`.